src/Cognito/Provider.php

```diff
@@ -0,0 +1,170 @@
+     */
+    protected $scopes = [
+        self::SCOPE_OPENID,
+        self::SCOPE_PROFILE,
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    protected $scopeSeparator = ' ';
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function additionalConfigKeys()
+    {
+        return [
+            'host',
+            'authorize_uri',
+            'token_uri',
+            'userinfo_uri',
+            'userinfo_key',
+            'user_id',
+            'user_nickname',
+            'user_name',
+            'user_email',
+            'user_avatar',
+        ];
+    }
+
+    /**
+     * Get the authentication URL for the provider.
+     *
+     * @param string $state
+     *
+     * @return string
+     */
+    protected function getAuthUrl($state)
+    {
+        return $this->buildAuthUrlFromBase($this->getCognitoUrl('authorize_uri'), $state);
+    }
+
+    /**
+     * Get the token URL for the provider.
+     *
+     * @return string
+     */
+    protected function getTokenUrl()
+    {
+        return $this->getCognitoUrl('token_uri');
+    }
+
+    /**
+     * Get the raw user for the given access token.
+     *
+     * @param string $token
+     *
+     * @return array
+     */
+    protected function getUserByToken($token)
+    {
+        $response = $this->getHttpClient()->get($this->getCognitoUrl('userinfo_uri'), [
+            'headers' => [
+                'Authorization' => 'Bearer '.$token,
+            ],
+        ]);
+
+        return (array) json_decode($response->getBody(), true);
+    }
+
+    /**
+     * Map the raw user array to a Socialite User instance.
+     *
+     * @param array $user
+     *
+     * @return \Laravel\Socialite\User
+     */
+    protected function mapUserToObject(array $user)
+    {
+        $key = $this->getConfig('userinfo_key', null);
+        $data = is_null($key) === true ? $user : Arr::get($user, $key, []);
+
+        return (new User())->setRaw($data)->map([
+            'id'       => $this->getUserData($data, 'sub'),
+            'nickname' => $this->getUserData($data, 'nickname'),
+            'name'     => $this->getUserData($data, 'name'),
+            'email'    => $this->getUserData($data, 'email'),
+            'avatar'   => $this->getUserData($data, 'picture'),
+
+            // user - array containing all available user attributes
+            // Cognito adheres to OIDC standard claims - https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+        ]);
+    }
+
+    /**
+     * Get the POST fields for the token request.
+     *
+     * @param string $code
+     *
+     * @return array
+     */
+    protected function getTokenFields($code)
+    {
+        return array_merge(parent::getTokenFields($code), [
+            'grant_type' => 'authorization_code',
+        ]);
+    }
+
+    protected function getCognitoUrl($type)
+    {
+        return rtrim($this->getConfig('host'), '/').'/'.ltrim(($this->getConfig($type, Arr::get([
```

Comment:
This statement is a little confusing, could we split it out into a couple of statements with a sprintf to concat together?

Reply:
done.

Reply:
Thanks for the feedback :)